crates/ruff/src/rules/pylint/rules/eq_without_hash.rs

```diff
@@ -0,0 +1,75 @@
+use rustpython_parser::ast::{self, Ranged, Stmt};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks that a class that implements a `__eq__` also implements `__hash__`.
+///
+/// ## Why is this bad?
+///
+///
+/// ## Example
+/// ```python
+/// class Person:
+///     def __init__(self):
+///         self.name = "monty"
+///
+///     def __eq__(self, other):
+///         return isinstance(other, Person) and other.name == self.name
+/// ```
+///
+/// Use instead:
+/// ```python
+/// class Person:
+///     def __init__(self):
+///         self.name = "monty"
+///
+///     def __eq__(self, other):
+///         return isinstance(other, Person) and other.name == self.name
+///
+///     def __hash__(self):
+///         return hash(self.name)
+/// ```
+#[violation]
+pub struct EqWithoutHash;
+
+impl Violation for EqWithoutHash {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Object does not implement `__hash__` method")
+    }
+}
+
+/// W1641
+pub(crate) fn object_without_hash_method(
+    checker: &mut Checker,
+    ast::StmtClassDef { name, body, .. }: &ast::StmtClassDef,
+) {
+    if !hash_eq_without_hash(body) {
+        return;
+    }
+    checker
+        .diagnostics
+        .push(Diagnostic::new(EqWithoutHash, name.range()));
+}
+
+fn hash_eq_without_hash(body: &[Stmt]) -> bool {
+    let mut has_hash = false;
+    let mut has_eq = false;
+
+    for statement in body {
+        let Stmt::FunctionDef(ast::StmtFunctionDef { name, .. }) = statement else {
+            continue;
+        };
+        if name == "__hash__" {
+            has_hash = true;
+        }
+        if name == "__eq__" {
+            has_eq = true;
+        }
+    }
```

Comment:
I wonder if it would make sense performance-wise to early-exit if we've seen both methods, which should be the case for code that has no violation. 

Up: Potentially fewer iterations
Down: More branching in a hot loop, may result in more branch misses.